### PR TITLE
[WordSeg] Reduce learning rate on plateau

### PR DESCRIPTION
**Pull request by texasmichelle**
Automatically reduce the learning rate when training loss stagnates.

Move training termination code into a function to reduce clutter in the main loop.

### CODE REVIEW

File: Examples/WordSeg/main.swift [hunk 1]
```diff
@@ -163,3 +164,36 @@ func usage() -> Never {
...
+  lossHistory: [Float], optimizer: Adam<SNLM>,
+  factor: Float = 0.25
+) {
+  let patience = 1
+  if lossHistory.count <= patience { return }
+  guard let loss = lossHistory.last else { return }
+
+  let window = Array(lossHistory.suffix(patience + 1))
+
+  if window.min() == loss && window.max() != loss { return }
```
Comment (sgugger) [inline Examples/WordSeg/main.swift:196]: The PyTorch implementation uses a threshold of 1e-4 to determine if there is an improvement: a is considered a better loss than b if a < b * (1-threshold) so here, there should be a similar test between `window.max()` and `loss` (probably `&& window.max() * (1-threshold) > loss`).

File: Examples/WordSeg/main.swift [hunk 2]
```diff
@@ -163,3 +164,36 @@ func usage() -> Never {
...
+
+  optimizer.learningRate = max(0.0, optimizer.learningRate * factor)
```
Comment (sgugger) [inline Examples/WordSeg/main.swift:198]: The PyTorch implementation has a minimal decay of 1e-8: if `optimizer.learningRate - optimizer.learningRate * factor <= 1e-8` nothing is done.

File: Examples/WordSeg/main.swift (hunk 1 above)
Comment (texasmichelle) [reply Examples/WordSeg/main.swift]: Done

File: Examples/WordSeg/main.swift (hunk 2 above)
Comment (texasmichelle) [reply Examples/WordSeg/main.swift]: Done

File: Examples/WordSeg/main.swift [hunk 3]
```diff
@@ -163,3 +164,41 @@ func usage() -> Never {
...
+  let minDecay: Float = 1e-8
+  if lossHistory.count <= patience { return }
+  guard let loss = lossHistory.last else { return }
+
+  let window = Array(lossHistory.suffix(patience + 1))
+
+  if loss >= window.min()! * (1 - threshold) && loss != window.max() { return }
+
+  let newLR = optimizer.learningRate * factor
+  if newLR > minDecay {
```
Comment (sgugger) [inline Examples/WordSeg/main.swift:201]: It should be `oldLR - newLR > minDecay` here (sorry if I was unclear before).

File: Examples/WordSeg/main.swift [hunk 4]
```diff
@@ -163,3 +164,41 @@ func usage() -> Never {
...
+) {
+  let patience = 1
+  let threshold: Float = 1e-4
+  let minDecay: Float = 1e-8
+  if lossHistory.count <= patience { return }
+  guard let loss = lossHistory.last else { return }
+
+  let window = Array(lossHistory.suffix(patience + 1))
+
+  if loss >= window.min()! * (1 - threshold) && loss != window.max() { return }
```
Comment (sgugger) [inline Examples/WordSeg/main.swift:198]: The first part of the test is always true: `window.min() ! * (1 - threshold) < window.min()! <= loss`.
The threshold is to consider there has been an improvement: we say newLoss is better than anything before if
`newLoss <= oldLoss * (1-threshold)` otherwise we consider there has not really been an improvement. 

Here we want to return if there has been an improvement during the window of time. I think that simply means `window.min()! <= window.max() * (1-threshold)` (also not sure why min needs a ! but not max?) If that test is false, it means all the loss in the window did not significantly improve compared to the threshold.

Comment (texasmichelle) [reply Examples/WordSeg/main.swift]: In this case, I believe `window.max()` does not require `!` because it is valid to compare an optional, whereby `window.min()` requires unwrapping to be used with `*`.

I attempted to clarify this function by assigning current and previous loss explicitly, rather than using min and max over a window (which was originally intended to support variable window sizes/patience values, but added too much complexity).
